sigstore/_internal/oidc.py

```diff
@@ -54,3 +66,177 @@ def __init__(self, identity_token: str) -> None:
             )
 
         self.proof: str = str(identity_jwt.get(proof_claim))
+
+
+AUTH_SUCCESS_HTML = """
+<html>
+<title>Sigstore Auth</title>
+<body>
+<h1>Sigstore Auth Successful</h1>
+<p>You may now close this page.</p>
+</body>
+</html>
+"""
+
+
+class RedirectHandler(http.server.BaseHTTPRequestHandler):
+    def log_message(self, format, *args):
+        pass
+
+    def do_GET(self):
+        server = cast(RedirectServer, self.server)
+        r = urllib.parse.urlsplit(self.path)
+
+        # Handle auth response
+        if r.path == server.redirect_path:
+            self.send_response(200)
+            self.send_header("Content-Type", "text/html; charset=utf-8")
+            body = AUTH_SUCCESS_HTML.encode("utf-8")
+            self.send_header("Content-Length", str(len(body)))
+            self.end_headers()
+            self.wfile.write(body)
+            server.auth_response = urllib.parse.parse_qs(r.query)
+            return
+
+        # Any other request generates an auth request
+        url = server.auth_request()
+        self.send_response(302)
+        self.send_header("Location", url)
+        self.end_headers()
+
+
+OOB_REDIRECT_URI = "urn:ietf:wg:oauth:2.0:oob"
+
+
+class RedirectServer(http.server.HTTPServer):
+    def __init__(self):
+        super().__init__(("127.0.0.1", 0), RedirectHandler)
+        self.state = None
+        self.nonce = None
+        self.auth_response = None
+        self.__port: int = self.socket.getsockname()[1]
```

Comment:
Nit: Single underscore is more idiomatic:

```suggestion
        self._port: int = self.socket.getsockname()[1]
```